fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+            features = {}
+
+        # Here we assume that global and local preds are simply tensors
+        # TODO: Perhaps loosen this at a later date.
+        if not isinstance(global_preds, torch.Tensor):
+            global_preds = global_preds["prediction"]
+        if not isinstance(local_preds, torch.Tensor):
+            local_preds = local_preds["prediction"]
+
+        assert isinstance(global_preds, torch.Tensor)
+        assert isinstance(local_preds, torch.Tensor)
+
+        if self.mkmmd_loss_weight != 0:
+            if not isinstance(self.model, MoonModel) or not isinstance(self.init_global_model, MoonModel):
+                AssertionError(
+                    "To compute the MK-MMD loss, the client model and the init_global_model must be of type MoonModel."
+                )
+            _, init_global_features = self.init_global_model(input)
+            features.update({"init_global_features": init_global_features["features"]})
+
+        return {"global": global_preds, "local": local_preds}, features
+
+    def compute_loss_and_additional_losses(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        """
+        Computes the loss and any additional losses given predictions of the model and ground truth data.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name.
+            features (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            Tuple[torch.Tensor, Union[Dict[str, torch.Tensor], None]]; A tuple with:
+                - The tensor for the total loss
+                - A dictionary with `local_loss`, `global_loss`, `total_loss` and, based on client attributes set
+                from server config, also `mkmmd_loss`, `feature_l2_norm_loss` keys and their respective calculated
+                values.
+        """
+
+        # Compute global model vanilla loss
+        assert "global" in preds
+        global_loss = self.criterion(preds["global"], target)
+
+        # Compute local model loss + ditto constraint term
+        assert "local" in preds
+        local_loss = self.criterion(preds["local"], target)
+        total_loss = local_loss.clone()
+
+        additional_losses = {"local_loss": local_loss, "global_loss": global_loss}
+
+        if self.mkmmd_loss_weight != 0:
+            assert "init_global_features" in features
+            assert "features" in features
+            # Compute MK-MMD loss
+            mkmmd_loss = self.mkmmd_loss(features["features"], features["init_global_features"])
+            total_loss += self.mkmmd_loss_weight * mkmmd_loss
+            additional_losses["mkmmd_loss"] = mkmmd_loss
+            if self.feature_l2_norm:
```

Comment:
I think technically, this component is "independent" of the mkmmd loss weight That is, we don't want to nest this if statement. Rather, even if mkmmd_loss_weight is zero, we could theoretically have an l2 norm loss.

Reply:
But I thought of l2 norm as additional component of mkmmd loss, do we need to have it separate than mkmmd loss?

Reply:
I see what you mean, sure I can decouple it from mkmmd loss, so this scenario wouldn't happen